### ISSUE TITLE
pt model to an ONNX model, different result

### ISSUE THREAD

**Issue by GXDD666** [ultralytics/ultralytics#19513]
When using the official tool to convert a PyTorch (pt) model to an ONNX model, it is found that when inferring the same image using the pt model and the ONNX model, the results are inconsistent.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @GXDD666, thank you for bringing this to our attention 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com/) for detailed guidance, where you may find useful information about model conversion and supported export formats.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the behavior you're encountering. This should include:

1. The steps you followed to export the model (e.g., using `YOLO.export()` command).
2. The specific image(s) and input data used for inference.
3. Any discrepancies or errors observed in the results between the PyTorch and ONNX models.

If you haven't already, ensure you are using the latest `ultralytics` package and dependencies. You can upgrade with the following command:

```bash
pip install -U ultralytics
```

## Resources to Check Out

- 📚 For instructions on exporting models and supported formats, see our [Export Documentation](https://docs.ultralytics.com/modes/export/).
- 🔍 Revisit our [ONNX Export Guide](https://docs.ultralytics.com/modes/export/#onnx) to ensure the export process follows best practices.

## Environments

YOLO models are best tested in up-to-date environments where dependencies are correctly configured. For easy testing, consider running the model utilizing pre-configured environments:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

If you’d like support with testing, these environments are also a great starting point.

## Community & Support

Connect with the Ultralytics community for more insights and discussions:
- 🚦 For real-time help, join our [Discord](https://discord.com/invite/ultralytics).
- 📋 Ask questions or share ideas on our [Discourse forum](https://community.ultralytics.com/).
- 🌐 Join our growing [Subreddit community](https://reddit.com/r/Ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing, ensuring proper functionality across all platforms. For more information on tested modes and tasks, see [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/).

😊 Please note that this is an automated response to help guide you to the most relevant resources. An Ultralytics engineer will review your issue shortly for further assistance!

**Comment by Y-T-G** [comment]
It's probably different because of the minimum rectangle resizing in PyTorch.

https://github.com/ultralytics/ultralytics/issues/16339#issuecomment-2358362884

If you export with `dynamic=True`. It should be the same.

**Comment by GXDD666** [comment]
this is my code. I infer the image "bus.jpg".

------------------------------------------------------------------------------------------------------------------------------
from ultralytics import YOLO
import os
import torch

os.environ['CUDA_VISIBLE_DEVICES'] = "3"
device=torch.device("cuda:{}".format(0))

# Load a model
model = YOLO("pretrained_model/yolo11n.pt")

# Perform object detection on an image
results = model("ultralytics/assets/bus.jpg")
results[0].show()
# print("results[0].keypoints")
# print(results[0].keypoints)
print(results[0].boxes)

# Export the model to ONNX format
path = model.export(format="onnx", device=device, batch=1,opset=11)  # return path to exported model
print(path)

# Load the exported ONNX model
onnx_model = YOLO(path)

# Run inference
results1 = onnx_model("ultralytics/assets/bus.jpg")
results1[0].show()
# print(results1[0])
# print(results1[0].keypoints)
print(results1[0].boxes)


------------------------------------------------------------------------------------------------------------------------------
pt result：

cls: tensor([5., 0., 0., 0., 0.], device='cuda:0')
conf: tensor([0.9402, 0.8882, 0.8783, 0.8558, 0.6219], device='cuda:0')
data: tensor([[3.8327e+00, 2.2936e+02, 7.9619e+02, 7.2841e+02, 9.4015e-01, 5.0000e+00],
        [6.7102e+02, 3.9483e+02, 8.0981e+02, 8.7871e+02, 8.8822e-01, 0.0000e+00],
        [4.7405e+01, 3.9957e+02, 2.3930e+02, 9.0419e+02, 8.7825e-01, 0.0000e+00],
        [2.2306e+02, 4.0869e+02, 3.4447e+02, 8.6044e+02, 8.5577e-01, 0.0000e+00],
        [2.1726e-02, 5.5607e+02, 6.8885e+01, 8.7236e+02, 6.2192e-01, 0.0000e+00]], device='cuda:0')
id: None
is_track: False
orig_shape: (1080, 810)
shape: torch.Size([5, 6])
xywh: tensor([[400.0136, 478.8882, 792.3619, 499.0481],
        [740.4135, 636.7728, 138.7925, 483.8793],
        [143.3527, 651.8801, 191.8959, 504.6298],
        [283.7633, 634.5622, 121.4086, 451.7470],
        [ 34.4536, 714.2139,  68.8638, 316.2907]], device='cuda:0')
xywhn: tensor([[0.4938, 0.4434, 0.9782, 0.4621],
        [0.9141, 0.5896, 0.1713, 0.4480],
        [0.1770, 0.6036, 0.2369, 0.4672],
        [0.3503, 0.5876, 0.1499, 0.4183],
        [0.0425, 0.6613, 0.0850, 0.2929]], device='cuda:0')
xyxy: tensor([[3.8327e+00, 2.2936e+02, 7.9619e+02, 7.2841e+02],
        [6.7102e+02, 3.9483e+02, 8.0981e+02, 8.7871e+02],
        [4.7405e+01, 3.9957e+02, 2.3930e+02, 9.0419e+02],
        [2.2306e+02, 4.0869e+02, 3.4447e+02, 8.6044e+02],
        [2.1726e-02, 5.5607e+02, 6.8885e+01, 8.7236e+02]], device='cuda:0')
xyxyn: tensor([[4.7318e-03, 2.1237e-01, 9.8296e-01, 6.7446e-01],
        [8.2842e-01, 3.6559e-01, 9.9977e-01, 8.1362e-01],
        [5.8524e-02, 3.6997e-01, 2.9543e-01, 8.3722e-01],
        [2.7538e-01, 3.7842e-01, 4.2527e-01, 7.9670e-01],
        [2.6822e-05, 5.1488e-01, 8.5044e-02, 8.0774e-01]], device='cuda:0')
------------------------------------------------------------------------------------------------------------------------------
onnx result：
cls: tensor([5., 0., 0., 0., 0.])
conf: tensor([0.9392, 0.9020, 0.8493, 0.8328, 0.3993])
data: tensor([[1.1918e+01, 2.2839e+02, 7.9923e+02, 7.3522e+02, 9.3920e-01, 5.0000e+00],
        [4.8593e+01, 3.9796e+02, 2.4321e+02, 9.0454e+02, 9.0199e-01, 0.0000e+00],
        [6.7056e+02, 3.9259e+02, 8.1000e+02, 8.7962e+02, 8.4928e-01, 0.0000e+00],
        [2.2312e+02, 4.0558e+02, 3.4523e+02, 8.5972e+02, 8.3279e-01, 0.0000e+00],
        [0.0000e+00, 5.5019e+02, 6.6008e+01, 8.7178e+02, 3.9934e-01, 0.0000e+00]])
id: None
is_track: False
orig_shape: (1080, 810)
shape: torch.Size([5, 6])
xywh: tensor([[405.5718, 481.8046, 787.3066, 506.8208],
        [145.9027, 651.2482, 194.6200, 506.5749],
        [740.2782, 636.1068, 139.4436, 487.0331],
        [284.1743, 632.6500, 122.1094, 454.1441],
        [ 33.0041, 710.9857,  66.0082, 321.5859]])
xywhn: tensor([[0.5007, 0.4461, 0.9720, 0.4693],
        [0.1801, 0.6030, 0.2403, 0.4691],
        [0.9139, 0.5890, 0.1722, 0.4510],
        [0.3508, 0.5858, 0.1508, 0.4205],
        [0.0407, 0.6583, 0.0815, 0.2978]])
xyxy: tensor([[ 11.9185, 228.3942, 799.2251, 735.2150],
        [ 48.5927, 397.9608, 243.2127, 904.5357],
        [670.5564, 392.5902, 810.0000, 879.6234],
        [223.1196, 405.5779, 345.2290, 859.7220],
        [  0.0000, 550.1927,  66.0082, 871.7786]])
xyxyn: tensor([[0.0147, 0.2115, 0.9867, 0.6808],
        [0.0600, 0.3685, 0.3003, 0.8375],
        [0.8278, 0.3635, 1.0000, 0.8145],
        [0.2755, 0.3755, 0.4262, 0.7960],
        [0.0000, 0.5094, 0.0815, 0.8072]])

**Comment by GXDD666** [comment]
> It's probably different because of the minimum rectangle resizing in PyTorch.
> 
> [#16339 (comment)](https://github.com/ultralytics/ultralytics/issues/16339#issuecomment-2358362884)
> 
> If you export with `dynamic=True`. It should be the same.

thx，it is work

**Comment by glenn-jocher** [comment]
Great to hear the `dynamic=True` export resolved your inference consistency! For optimal ONNX exports, ensure you're using the latest `ultralytics` package and consider omitting explicit device assignment since export handles this automatically. The Export Mode documentation provides additional guidance on preserving model behavior across formats.

**Comment by glenn-jocher** [comment]
For YOLO11 inference consistency between PyTorch and ONNX, ensure you're using identical preprocessing and let the model handle it automatically. A minimal example:

```python
from ultralytics import YOLO
model = YOLO('yolo11n.pt')
results = model.predict(source='image.jpg')  # PT inference
model.export(format='onnx', dynamic=True, opset=12)  # Export
onnx_results = YOLO('yolo11n.onnx').predict('image.jpg')  # ONNX inference
```

Key checks:
1. Update to latest `ultralytics>=8.2.11`
2. Avoid manual preprocessing - the `YOLO()` class handles normalization/resizing
3. Compare raw outputs with `results[0].boxes.data` vs `onnx_results[0].boxes.data`

If discrepancies persist, please create a [Minimum Reproducible Example](https://docs.ultralytics.com/guides/yolo-common-issues/#minimum-reproducible-example) using official YOLO11n weights and share via GitHub Issues.
